### PR TITLE
MNT: Run PTH flake test in prep for supporting pathlib (io.ascii)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Ref #16924
This is in the continuation of #16060

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/io/ascii/ui.py [hunk 1]
```diff
@@ -1059,10 +1058,14 @@ def write(
         # behavior is for Python to translate \r\n (which we write because
         # of os.linesep) into \r\r\n. Specifying newline='' disables any
         # auto-translation.
-        output = open(output, "w", newline="")
-        output.write(outstr)
-        output.write(os.linesep)
-        output.close()
+        if not output:
```
Comment (pllim) [inline astropy/io/ascii/ui.py:1061]: But there is already this check above:

```python
    if output is None:
        output = sys.stdout
```

Comment (neutrinoceros) [reply astropy/io/ascii/ui.py:1061]: This check is different: it also excludes the empty string `''` because `Path('')` means "the current work directory", and this scenario (passing `output=''` and expecting a `FileNotFoundError`) is actually exercised in a test.

Comment (pllim) [reply astropy/io/ascii/ui.py:1061]: What was the old behavior for `output=''`? Does this need a change log?

Comment (neutrinoceros) [reply astropy/io/ascii/ui.py:1061]: Au contraire, this is being strictly backward compatible: `open('')` raises the same exception.

File: astropy/io/ascii/qdp.py [hunk 2]
```diff
@@ -136,9 +138,8 @@ def _get_type_from_list_of_lines(lines, delimiter=None):
 def _get_lines_from_file(qdp_file):
     if "\n" in qdp_file:
         lines = qdp_file.split("\n")
-    elif isinstance(qdp_file, str):
-        with open(qdp_file) as fobj:
-            lines = [line.strip() for line in fobj]
+    elif isinstance(qdp_file, (str, os.PathLike)):
+        lines = [line.strip() for line in Path(qdp_file).read_text().splitlines()]
```
Comment (taldcroft) [inline astropy/io/ascii/qdp.py:142]: This new code ends up making two additional copies of the entire file in memory, one the full `read_text()` output and second the `splitlines()` output. In general `splitlines()` is not a drop-in replacement for iterating over `open()` since `splitlines()` removes the line endings. In this case it doesn't matter but best to keep this `Path` refactoring as close to the original code as possible.

So basically the only change here should be `isinstance(qdp_file, (str, os.PathLike)):`.

File: astropy/io/ascii/ui.py (hunk 1 above)
Comment (taldcroft) [reply astropy/io/ascii/ui.py:1061]: I am seeing this error message:
```
>>> open("", "w")
FileNotFoundError: [Errno 2] No such file or directory: ''
```

File: astropy/io/ascii/ui.py [hunk 3]
```diff
@@ -1005,8 +1004,8 @@ def write(
         "write", format=format, fast_writer=fast_writer, overwrite=overwrite, **kwargs
     )
 
-    if isinstance(output, (str, bytes, os.PathLike)):
-        output = os.path.expanduser(output)
+    if isinstance(output, (str, bytes, os.PathLike)) and output:
+        output = Path(output).expanduser()
```
Comment (taldcroft) [inline astropy/io/ascii/ui.py:1008]: I would prefer to revert the changes In this function and stick with the original `os.path.expanduser()`. In this case using `Path` is introducing new complexity to the logic because of the `Path("") => cwd()` behavior. So maybe add a `#noqa ...`.

File: astropy/io/ascii/ui.py [hunk 4]
```diff
@@ -1059,10 +1058,14 @@ def write(
...
-        output = open(output, "w", newline="")
-        output.write(outstr)
-        output.write(os.linesep)
-        output.close()
+        if not output:
+            # validate that output is non-empty, as
+            # Path("") gives the current working directory
+            raise FileNotFoundError("No such file or directory")
+        Path(output).write_text(
+            outstr + os.linesep,
```
Comment (taldcroft) [inline astropy/io/ascii/ui.py:1066]: This creates an unnecessary temporary string `outstr + os.linesep`.

File: astropy/io/ascii/ui.py [hunk 5]
```diff
@@ -1059,10 +1058,14 @@ def write(
         # behavior is for Python to translate \r\n (which we write because
         # of os.linesep) into \r\r\n. Specifying newline='' disables any
         # auto-translation.
-        output = open(output, "w", newline="")
-        output.write(outstr)
-        output.write(os.linesep)
-        output.close()
```
Comment (taldcroft) [inline astropy/io/ascii/ui.py:1065]: This should just be changed to an `open(...)` context manager.

**Review by taldcroft — CHANGES_REQUESTED** [review]
Looks OK but with requested changes. Overall in the `Path` migration I think you should mostly avoid doing a blind replacement of `open(...)` with `Path().read/write_text()`. 

There is nothing wrong with `open()` and the two cases here show that `open()` is often preferable because it is a streaming interface.

File: astropy/io/ascii/cds.py [hunk 6]
```diff
@@ -379,7 +380,7 @@ def read(self, table):
             with suppress(TypeError):
                 # For strings only
                 if os.linesep not in table + "":
```
Comment (astrofrog) [inline astropy/io/ascii/cds.py:382]: @taldcroft just out of pure curiosity, what is the purpose of the + "" here?

**Comment by neutrinoceros** [comment]
>  Overall in the Path migration I think you should mostly avoid doing a blind replacement of open(...) with Path().read/write_text().
>
> There is nothing wrong with open() and the two cases here show that open() is often preferable because it is a streaming interface.

@taldcroft I agree. I'll open a PR to propose ignoring [that specific rule (`PTH123`)](https://docs.astral.sh/ruff/rules/builtin-open/) more permanently. Moving this one to draft until it is resolved.

File: astropy/io/ascii/cds.py (hunk 6 above)
Comment (neutrinoceros) [reply astropy/io/ascii/cds.py:382]: I assume this is to trigger a `TypeError` if `table` isn't a string ?

Comment (taldcroft) [reply astropy/io/ascii/cds.py:382]: Exactly what @neutrinoceros said. This is a very old idiom to check for string-like (somewhat like a protocol, asking if it supports appending a `str`) instead of an explicit `str` type check.

What's the best way to check for string-like? Or does `isinstance(table, str)` cover all realistic use cases?

Comment (dhomeier) [reply astropy/io/ascii/cds.py:382]: From `BaseReader.read` in core.py below, which uses the same construct, the only other inputs are file-like or list of strings (the table lines), so they should all be covered.

File: astropy/io/ascii/ui.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/io/ascii/ui.py:1061]: I'm going to revert this and simply refactor with a context manager as suggested in https://github.com/astropy/astropy/pull/16954/files#r1748045678

File: astropy/io/ascii/qdp.py (hunk 2 above)
Comment (neutrinoceros) [reply astropy/io/ascii/qdp.py:142]: > In this case it doesn't matter but best to keep this Path refactoring as close to the original code as possible.

100% agreed. Any behavioral difference with the original code that I don't advertise myself in a comment is a mistake from me, so thank you for your attention to details !

File: astropy/io/ascii/ui.py (hunk 5 above)
Comment (neutrinoceros) [reply astropy/io/ascii/ui.py:1065]: done

**Comment by neutrinoceros** [comment]
I think I've now addressed all comments from @taldcroft, let me know if there's anything I missed !

File: astropy/io/ascii/ui.py [hunk 7]
```diff
@@ -1059,10 +1058,10 @@ def write(
...
+        with open(output, "w", newline="") as output:
+            output.write(outstr)
+            output.write(os.linesep)
+            output.close()
```
Comment (taldcroft) [inline astropy/io/ascii/ui.py:1064]: The `.close()` is not needed with the context manager.

Comment (neutrinoceros) [reply astropy/io/ascii/ui.py]: Thanks for catching this. Fixed.

**Review by taldcroft — APPROVED** [review]
